### PR TITLE
typed interfaces

### PR DESCRIPTION
**Pull request by frairon**
- update go dependencies, update to go 1.21
- typed interfaces initial

### CODE REVIEW

File: gtyped/codecs.go [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+type GCodec[V any] interface {
+	GEncode(value V) (data []byte, err error)
+	GDecode(data []byte) (value V, err error)
+}
+
+type codecBridge[T any] struct {
+	c GCodec[T]
+}
+
+func (cb *codecBridge[T]) Encode(value interface{}) ([]byte, error) {
```
Comment (paulhenke) [inline gtyped/codecs.go]:
```suggestion
func (cb *codecBridge[T]) Encode(value any) ([]byte, error) {
```

File: gtyped/codecs.go [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+	tVal, ok := value.(T)
+	var zero T
+	if !ok {
+		return nil, fmt.Errorf("unexpected type while encoding. Expected %T, got %T", zero, value)
+	}
+	return cb.c.GEncode(tVal)
+}
+
+func (cb *codecBridge[T]) Decode(data []byte) (interface{}, error) {
```
Comment (paulhenke) [inline gtyped/codecs.go]:
```suggestion
func (cb *codecBridge[T]) Decode(data []byte) (any, error) {
```

File: gtyped/codecs.go [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+	"github.com/lovoo/goka"
+)
+
+type GCodec[V any] interface {
+	GEncode(value V) (data []byte, err error)
+	GDecode(data []byte) (value V, err error)
+}
+
+type codecBridge[T any] struct {
```
Comment (paulhenke) [inline gtyped/codecs.go]: The GOF name is adapter. 
So maybe?
```suggestion
type codecAdapter[T any] struct {
```

Also with the ctor func below.

File: gtyped/codecs.go [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+		return nil, fmt.Errorf("unexpected type while encoding. Expected %T, got %T", zero, value)
+	}
+	return cb.c.GEncode(tVal)
+}
+
+func (cb *codecBridge[T]) Decode(data []byte) (interface{}, error) {
+	return cb.c.GDecode(data)
+}
+
+// CodecBridge converts a untyped (goka) codec to a typed codec
```
Comment (paulhenke) [inline gtyped/codecs.go]: From the signature, it takes a typed codec and returns the untyped codec.
The comment is the other way around?

File: gtyped/codecs.go [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+func NewCodecBridge[T any](codec GCodec[T]) goka.Codec {
+	return &codecBridge[T]{
+		c: codec,
+	}
+}
+
+type StringCodec[T ~string] struct{}
+
+func (s *StringCodec[T]) GEncode(value T) (data []byte, err error) {
```
Comment (paulhenke) [inline gtyped/codecs.go]:
```suggestion
func (s *StringCodec[T]) GEncode(value T) ([]byte, error) {
```

File: gtyped/codecs.go (hunk 1 above)
Comment (frairon) [reply gtyped/codecs.go]: done, thanks

File: gtyped/codecs.go (hunk 3 above)
Comment (frairon) [reply gtyped/codecs.go]: absolutely, yeah why not. let's rename it.

File: gtyped/codecs.go (hunk 4 above)
Comment (frairon) [reply gtyped/codecs.go]: true that, changed the comment.

File: gtyped/codecs.go (hunk 5 above)
Comment (frairon) [reply gtyped/codecs.go]: stupid copy paste :). Thanks

**Review by jomaresch — APPROVED** [review]

**Pull request merged**